### PR TITLE
Refactor versionbits deployments to avoid potential uninitialized variables

### PR DESCRIPTION
**Pull request by achow101**
Currently versionbits deployments are stored in a simple C array in `Params`. Setting the actual values needs to be done externally. Since the memory is allocated and initialized when `Params` is initialized, it is possible to have a deployment with uninitialized (or otherwise garbage and invalid) values set for its parameters if something is forgotten when they are being set.

Instead of this two step approach (create deployment object, then set its members), I think it is better to have a constructor where those parameters are passed in and set when the deployment is initialized. This PR does that. In order for this to work though, `vDeployments` cannot be a simple C array. It is instead changed to be a `std::map<DeploymentPos, VBitsDeployment>`. This maps the deployment enum to the `VBitsDeployment` object itself; this is essentially how we were using `vDeployments` previously. Additionally, by also changing `VersionBitsDeploymentInfo` to a map as well, we are able to remove `MAX_VERSION_BITS_DEPLOYMENT` and all of the places where we were iterating the deployments no longer need to cast an int to `DeploymentPos`.

An additional concern is that the members of `VBitsDeployment` are not const and could thus be accidentally changed during runtime. To prevent this, this PR makes all members of `VBitsDeployment` const.

Lastly, to make `VBitsDeployment` actually fully describe a deployment, this PR moves the number of blocks in a period into the struct. In order to ensure that this variable matches up with the retarget period, a unit test is added.

Built on top of #21392 and #21380 for the refactors and other changes they make to `VBitsDeployment`.

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -340,18 +324,10 @@ class SigNetParams : public CChainParams {
...
-        // Activation of Taproot (BIPs 340-342)
-        consensus.m_deployments[Consensus::DEPLOYMENT_TAPROOT].bit = 2;
-        consensus.m_deployments[Consensus::DEPLOYMENT_TAPROOT].startheight = Consensus::VBitsDeployment::ALWAYS_ACTIVE;
-        consensus.m_deployments[Consensus::DEPLOYMENT_TAPROOT].timeoutheight = Consensus::VBitsDeployment::NO_TIMEOUT;
-        consensus.m_deployments[Consensus::DEPLOYMENT_TAPROOT].threshold = 1916; // 95% of 2016
-        consensus.m_deployments[Consensus::DEPLOYMENT_TAPROOT].m_min_activation_height = 0; // No minimum activation height
+        consensus.m_deployments.emplace(Consensus::DEPLOYMENT_TESTDUMMY, Consensus::VBitsDeployment(28, false /* active */));
+
+        // Deployment of Taproot (BIPs 340-342)
+        consensus.m_deployments.emplace(Consensus::DEPLOYMENT_TAPROOT, Consensus::VBitsDeployment(2, true /* active */));
```
Comment (ajtowns) [inline src/chainparams.cpp:330]: We usually do these comments as `/* active= */ false` I think

**Review by ajtowns — COMMENTED** [review]
`globalChainParams` / `Params()` is already const during runtime, so that should already prevent changing any of the fields in `vDeployments`. Ah, a different approach that might be worth considering:

```c++
class CChainParams
{
protected:
    CChainParams(const Consensus::Params& cons) : consensus{cons} { } // or std::move, whatever
public:
    const Consensus::Params consensus;
    ...
};

class CRegTestParams : public CChainParams
{
public:
    CRegTestParams(const ArgsManager& args) : CChainParams{getConsensus(args)}
    { ... }
    
    static Consensus::Params getConsensus(const ArgsManager& args)
    {
        Consensus::Params consensus{};
        consensus.foo = bar; ...
        // parse -segwitheight and -vbparams here as well
    }
};
```

and have default initializers for everything in `Consensus::Params`?

Not keen on the map -- the possibility that `deployments.at(DEPLOYMENT_FOO)` might throw doesn't seem that great to me; and if you're just using `deployments[DEPLOYMENT_FOO]` and relying on it getting default initialised to something sane, might as well give VBitsDeployment a default initiaizer and just use the array?

**Comment by achow101** [comment]
> Ah, a different approach that might be worth considering:

That doesn't seem that much different from what we already do now.

> Not keen on the map -- the possibility that `deployments.at(DEPLOYMENT_FOO)` might throw doesn't seem that great to me; and if you're just using `deployments[DEPLOYMENT_FOO]` and relying on it getting default initialised to something sane, might as well give VBitsDeployment a default initiaizer and just use the array?

It seems better to me to throw (and hard shutdown) than to use some defaults, or worse, uninitialized members. I don't really like the idea of consensus params being initialized with default values that might not actually be what is wanted.

**Comment by practicalswift** [comment]
Concept ACK on not using error prone C arrays like maps.

**Comment by jnewbery** [comment]
This is built on #21392, which is now closed. Should this PR also be closed?

**Comment by MarcoFalke** [comment]
I'd say it needs rebase after #21377 is merged

**Comment by luke-jr** [comment]
This seems to be a lot less readable. Is there a good way to get uninitialised warnings without harming readability, perhaps?

**Comment by achow101** [comment]
Closing for now, may revisit in the future, or someone else can take it over.

**Pull request closed without merging**